Sort maven project list in Explorer and hide invalid pom.xml files

owenconti opened the pull request:
- Fixed a TS typing error on a boolean variable.
- Updated `_ensureMavenProjectParsed` of `MavenProjectNode` to become public.
- Alphabetically sort the array of `MavenProjectNode` items in `WorkspaceFolderNode`.

<img width="376" alt="screen shot 2018-08-31 at 6 21 31 pm" src="https://user-images.githubusercontent.com/791222/44940390-08683680-ad4b-11e8-8378-bb4583ffd030.png">

If I've made any mistakes or if there is a better way of approaching the problem, please let me know.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vscode-maven?pullRequest=118) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vscode-maven?pullRequest=118) before we can accept your contribution.<br/><br/>:x: owenconti [sign now](https://cla.opensource.microsoft.com/Microsoft/vscode-maven?pullRequest=118)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vscode-maven?pullRequest=118) it.</sub>

owenconti commented:
Hi @Eskibear,

I've updated the code to handle errors thrown in the `ensureMavenProjectParsed` function. I also updated `getChildren` in `WorkspaceFolderNode` to make sure the `MavenProjectNode` is valid before considering it a children. 

Here is a screenshot showing the difference:

<img width="757" alt="screen shot 2018-09-01 at 4 37 06 pm" src="https://user-images.githubusercontent.com/791222/44950461-b6d0b200-ae05-11e8-95d6-8781613eeab8.png">

Valid pom.xml ^

<img width="779" alt="screen shot 2018-09-01 at 4 37 16 pm" src="https://user-images.githubusercontent.com/791222/44950462-c05a1a00-ae05-11e8-8b97-32ce83a1545f.png">

Invalid pom.xml ^ (notice the project missing from the Explorer list)

Eskibear commented on src/explorer/model/MavenProjectNode.ts:
```diff
@@ -49,14 +50,24 @@ export class MavenProjectNode extends NodeBase {
     }
 
-    private async _ensureMavenProjectParsed(): Promise<void> {
+    public async ensureMavenProjectParsed(): Promise<void> {
```
`ensureMavenProjectParsed` actually doesn't ensure a maven project is successfully parsed, `isValidPom` should always be checked after calling this function. So I think it's better to make this private, and only to expose `isValidPom`.

Eskibear commented on src/explorer/model/MavenProjectNode.ts:
```diff
@@ -49,14 +50,24 @@ export class MavenProjectNode extends NodeBase {
         return this._mavenProject;
     }
+
+    public async isValidPom(): Promise<boolean> {
```
the class is `MavenProjectNode`, so `hasValidPom` might be a better name.

Eskibear commented on src/explorer/model/MavenProjectNode.ts:
```diff
@@ -18,7 +18,8 @@ export class MavenProjectNode extends NodeBase {
 
     public async getTreeItem(): Promise<vscode.TreeItem> {
-        await this._ensureMavenProjectParsed();
+        await this.ensureMavenProjectParsed();
```
There is a potential NPE for `this._mavenProject.name` in next line. Although currently in this PR projects with invalid pom have been filtered out in `WorkspaceFolder.getChildren`, it's safer to check it here to avoid potential NPEs in future.

e.g. something like
```typescript
if (! await this.isValidPom()) {
    return undefined;
} 
```

Eskibear commented on src/explorer/model/WorkspaceFolderNode.ts:
```diff
@@ -39,4 +49,10 @@ export class WorkspaceFolderNode extends NodeBase {
         this._pomPaths = await Utils.getAllPomPaths(this._workspaceFolder);
     }
+
+    private async _sortChildren(): Promise<void> {
```
I don't see any async things inside, so I think there's no need to declare as an `async` function.

Eskibear commented on src/explorer/model/WorkspaceFolderNode.ts:
```diff
@@ -28,7 +28,17 @@ export class WorkspaceFolderNode extends NodeBase {
+            }
+        }
+
+        await this._sortChildren();
```
no need to use async/await for the _sortChildren function.

owenconti commented:
I think I got all the changes. I also changed `_ensureMavenProjectParsed` to `_parseMavenProject` to more accurately describe what it does.

Apologies, I hadn't seen that you made an issue for the parsing of invalid pom.xml files.
